### ISSUE TITLE
Command 'apt-get update' returned non-zero exit status 100

### ISSUE THREAD

**Issue by astroboyhere** [calamares/calamares#2213]
Occasionally and for reasons still unknown to us, in our distro (Escuelas Linux) Calamares fails to install with the message:

"Command 'apt-get update' returned non-zero exit status 100."

This message seems to appear only on UEFI machines, and even more strangely, on computers that do not have an internet connection, so there would be no reason to trigger an 'apt-get update'.

Given that our distribution does not require any use of the Internet during the installation process, I wonder if there is a line or lines of code that could be modified or suppressed to avoid triggering the apt-get update command in our distro installation.

P.S. We were not able to use #calamares:kde.org to ask for help in this regard, as the webchat announces that "Registration has been disabled on this homeserver."

**Comment by statex-official** [comment]
create account by this link app.element.io

**Comment by astroboyhere** [comment]
Thanks to the lead of statex-official, I was finally able to access Calamares Talk.

There I get a promising hint from ade: set update_db to false in packages.conf. I will try it and report back in a few days when we return to a school that has five computers with this problem.

Using a search engine with the word Calamares and the title of this issue, it seems that many users have also found this issue, and even describe the same environment as we do: UEFI and not connected to the Internet. No one had found a solution AFAIK.

Nevertheless, it would be difficult to find this bug in the Calamares code, as it happens occasionally on some computers for unknown reasons.

**Comment by kkofler** [comment]
As I understand it, the problem is that Debian in their infinite wisdom made the GRUB 2 BIOS and GRUB 2 UEFI packages mutually exclusive, so the Calamares configuration for Debian wants to replace the GRUB 2 BIOS package with the UEFI version on UEFI systems. The Debian GRUB packaging needs to be fixed. All other distros allow having both the BIOS and UEFI versions of GRUB 2 installed together. Only the one that is actually needed will be installed to the bootloader (by `grub-install` or by whatever mechanism the distribution uses).

**Comment by astroboyhere** [comment]
The issue we described here occasionally happens in the 64-bit edition of our distro, which uses an Ubuntu 22.04 base. It has the grub packages for BIOS and UEFI installed and coexisting peacefully (grub-efi… grub-pc…)

We also maintain a 32-bit edition based on Debian 11, but we have no issues with Calamares in that edition. If Debian is working fine here, their policies regarding grub are not the source of our issue, at least for our case, as we would not endorse installing a 32-bit Debian on a UEFI computer.

As I understand, you thought we were using Debian for our 64-bit edition, so it would try to download the non-available grub UEFI packages. To do so, it would need to run an 'apt-get update' to be able to fetch the missing grub EFI packages, right?

Although it was not our case because we do not use Debian on 64-bit, the source issue is that Calamares sometimes thinks it is on an Internet-connected computer. Hence, it triggers an 'apt-get update,' and crashes the installer because there is no Internet.

Setting update_db to false in packages.conf could be a working workaround, but we still wouldn't know why an 'apt-get update' is sometimes triggered on computers without an Internet connection.

**Comment by astroboyhere** [comment]
Well, I'm reopening the thread because I came across another computer with this problem (a Lenovo C260, by the way).

Although I set `update_db=false` in packages.conf, the Calamares installer crashed with the message "Command 'apt-get update' returned non-zero exit status 100". Oddly enough, the computer was not connected to the internet, so there would be no reason to run "apt-get update". There were 4 other UEFI computers of other brands in the school, also without internet connection, but they had no problems with the Calamares installer.

Since it would be very difficult to pinpoint a problem that occurs occasionally and for no apparent reason, I wonder if it would be possible to remove `apt-get update` from the Calamares code, since our distribution (Escuelas Linux) does not require an internet connection to install. If so, which line(s) and file(s) would have to be changed or removed to achieve this?

Thanks in advance.

**Comment by dalto8** [comment]
> I wonder if it would be possible to remove `apt-get update` from the Calamares code

If you really want to hack it out completely it is [here](https://github.com/calamares/calamares/blob/calamares/src/modules/packages/main.py#L274).

**Comment by adriaandegroot** [comment]
You can edit the `main.py` in your live system. While you're doing that, you might want to add something like

```
    libcalamares.utils.debug(str(libcalamares.job.configuration))
```

at the start of `run()` in the module, to check that your job configuration is actually what you think it is. This is the only location in the Calamares source that `apt-get update`  occurs. Since you have not provided a backtrace, there's nothing else for us to go on: it's not even certain that the `apt-get update` command is being run by Calamares, and a message like "Command 'apt-get update' returned non-zero exit status 100" is not produced by Calamares directly.

Searching carefully, it seems this was mentioned in a sparkylinux forum in 2019 -- and never resolved or reported upstream.

**Comment by astroboyhere** [comment]
Thanks for your answers guys!

In two days I will have access to the same machine, so I will reinstall on it to reproduce this problem again, 

On the live system, I will edit the main.py file in the packages directory to add the suggested libcalamares.utils.debug string.

As for the line:

```
    def update_db(self):
        check_target_env_call(["apt-get", "update"])
```

Should I replace it with:

```
    def update_db(self):
        pass
```

Or should I type something else to skip running apt-get update?

**Comment by astroboyhere** [comment]
I had access to the same machine and on the live USB stick I modified `/usr/lib/x86_64-linux-gnu/calamares/modules/packages/main.py` by replacing "`check_target_env_call(["apt-get", "update"])`" with "`pass`" (this without being sure if it should be done) and added "`libcalamares.utils.debug(str(libcalamares.job.configuration))`" on the line after `run()`. 

No luck, the Calamares installer still crashed with the message "_Command 'apt-get update' returned non-zero exit status 100"_. I was also not sure what `libcalamares.utils.debug` was supposed to be doing when I checked /var/log for a possible activity log.

Fortunately, the school was gracious enough to let me take their computer home, which saved me another long trip since it is located in a remote rural area.

If something is calling "_apt-get update_", it could be any file in `/etc`, right? So I grep that string and... Bingo! The culprit is `/etc/calamares/modules/before_bootloader_context.conf`, a file provided by a third party package: calamares-settings-lubuntu.

I then proceeded to comment out any reference to that file in `/etc/calamares/settings.conf`, as `before_bootloader_context.conf` does not seem to contain anything relevant for our distro. 

And it worked! The Calamares installer finished without any problems.

It might be interesting to know why` before_bootloader_context.conf `fails on a few machines without internet connection, while on most machines there are no problems at all, with or without Internet connection. But at least now some users of our distro would not have this problem, and maybe this solution could be useful for some other distros as well.
